target/linux/mediatek/patches-6.1/251-pinctrl-mediatek-mt7981-fix-emmc_45-group.patch

```diff
@@ -0,0 +1,34 @@
+mediatek: mt7981: pinctrl: fix emmc_45 group
+
+Add pin 0 with function 2 to the emmc_45 group in the pinctrl driver
+for the MediaTek MT7981 SoC.
+
+Signed-off-by: Jean Thomas <jean.thomas@wifirst.fr>
+---
+ drivers/pinctrl/mediatek/pinctrl-mt7981.c | 6 +++---
+ 1 file changed, 3 insertions(+), 3 deletions(-)
+
+--- a/drivers/pinctrl/mediatek/pinctrl-mt7981.c
++++ b/drivers/pinctrl/mediatek/pinctrl-mt7981.c
+@@ -700,8 +700,8 @@
+ static int mt7981_drv_vbus_funcs[] = { 1, };
+
+ /* EMMC */
+-static int mt7981_emmc_45_pins[] = { 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
+-static int mt7981_emmc_45_funcs[] = { 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, };
++static int mt7981_emmc_45_pins[] = { 0, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
```

Comment:
So now I took a look at the datasheet and checked what function 2 of GPIO 0 actually does: `WA_AICE_TCKC`. So that's a JTAG clock signal which is intended to be used when debugging the built-in wireless controller. And that smells a bit fishy.
I totally believe that this fixes the issue of the eMMC not working, but we have to find out why. Is GPIO 0 really used as a signal for the eMMC on that board? If so, do we really have to set function 2, or is it enough to drive the pin as output and set it either low or high using a GPIO hog?

Reply:
> So now I took a look at the datasheet and checked what function 2 of GPIO 0 actually does: WA_AICE_TCKC

Sadly I don't have access to the datasheet, and I didn't find it online. Would it be possible to email it to me, if it is not protected by a NDA?

Regarding the eMMC I initially added the GPIO 0 with function 2 because it is what is done in the GL.iNet firmware. However, leaving it unclaimed works on my device.  Maybe the pin is driven by the hardware with a pull-up or a pull-down resistor?

> or is it enough to drive the pin as output and set it either low or high using a GPIO hog?

You mean adding a GPIO hog in the device tree, just in case it is not really HW-driven?
```
&pio {
	gpio_wps-hog {
		gpio-hog;
		gpios = <0 GPIO_ACTIVE_HIGH>;
		output-high;
	};
        [...]
}
```

But there is another issue which will force me to create a new MMC group: the GPIO 15 present in the `emmc_45` group is also used by the `pwm0_1` [group](https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/files-5.15/drivers/pinctrl/mediatek/pinctrl-mt7981.c#L676-L677), which is used to drive the fan with the `pwm0_pin` [pin](https://github.com/JeThWifirst/openwrt/blob/add-gl-inet-xe3000-support/target/linux/mediatek/dts/mt7981a-glinet-gl-x3000-xe3000-common.dtsi#L195-L200).

So my plan is to rework my second pinctrl patch in order to create a new `mmc` group, without the pins 0 and 15:
```c
static int mt7981_emmc_pins[] = { 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
static int mt7981_emmc_funcs[] = { 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, };
```

Would that solution be okay for you?

Reply:
The last version I pushed implements this solution

Reply:
> So my plan is to rework my second pinctrl patch in order to create a new `mmc` group, without the pins 0 and 15:

So I've been stairing at the datasheet for a short while and concluded:

* Removing pin 15 can work because that's *EMMC_RSTB*, and maybe the reset of the eMMC just isn't connected or hard-wired to board reset circuit or another pin **in GPIO mode** is can be used to drive the reset signal. That liberates pin 15 of the SoC which can then be used for PWM instead, for example. 

* Setting pin 0 to function 2 doesn't make sense, that would be a the JTAG clock of the wireless MAC co-processor used when debugging the WA firmware using JTAG. If it is really used as reset for the emmc, it should be wired up as `reset-gpios` of the mmc bus in device tree.

* Pin 16~19 will always be used as *EMMC_DAT0*...*EMMC_DAT3*.

* Pin 20~23 are used in case we are using 8-bit bus-width for the eMMC, otherwise they can be liberated as well.

Hence, to me it would make most sense to have pin 15 in a separate group of it's own called emmc_reset, and have emmc4 and emmc8 groups for 4 bit and 8 bit bus-width respectively.

I hope I could give some insights here. Let me know if further assistance is needed so we can make this move forward.